compositor_pipeline/src/pipeline/input/mp4.rs

```diff
@@ -0,0 +1,301 @@
+                video_thread: Some(video_thread),
+                delete_file,
+                source_file: input_path,
+                stop_video_reader,
+            },
+            ChunksReceiver {
+                video: Some(video_receiver),
+                audio: None,
+            },
+        ))
+    }
+}
+
+impl Drop for Mp4 {
+    fn drop(&mut self) {
+        if let Some(thread) = self.video_thread.take() {
+            self.stop_video_reader
+                .store(true, std::sync::atomic::Ordering::Relaxed);
+            thread.join().unwrap();
+        }
+
+        if self.delete_file {
+            std::fs::remove_file(&self.source_file).unwrap();
+        }
+    }
+}
+
+fn spawn_video_reader(
+    reader: Mp4Reader<File>,
+    input_id: InputId,
+    stop_thread: Arc<AtomicBool>,
+) -> Result<(JoinHandle<()>, Receiver<EncodedChunk>), Mp4Error> {
+    let (&track_id, track, avc1) = reader
+        .tracks()
+        .iter()
+        .find_map(|(id, track)| {
+            let track_type = match track.track_type() {
+                Ok(t) => t,
+                Err(_) => return None,
+            };
+
+            let media_type = match track.media_type() {
+                Ok(t) => t,
+                Err(_) => return None,
+            };
+
+            if track_type != mp4::TrackType::Video
+                || media_type != mp4::MediaType::H264
+                || track.trak.mdia.minf.stbl.stsd.avc1.is_none()
+            {
+                return None;
+            }
+
+            track
+                .trak
+                .mdia
+                .minf
+                .stbl
+                .stsd
+                .avc1
+                .as_ref()
+                .map(|avc1| (id, track, avc1))
+        })
+        .ok_or(Mp4Error::NoTrack)?;
+
+    let (sender, receiver) = crossbeam_channel::bounded(10);
+
+    // sps and pps have to be extracted from the container, interleaved with [0, 0, 0, 1],
+    // concatenated and prepended to the first frame.
+    let sps = avc1
+        .avcc
+        .sequence_parameter_sets
+        .iter()
+        .flat_map(|s| [0, 0, 0, 1].iter().chain(s.bytes.iter()));
+
+    let pps = avc1
+        .avcc
+        .picture_parameter_sets
+        .iter()
+        .flat_map(|s| [0, 0, 0, 1].iter().chain(s.bytes.iter()));
+
+    let sps_and_pps_payload = Some(sps.chain(pps).copied().collect::<Bytes>());
+
+    let sample_count = track.sample_count();
+    let timescale = track.timescale();
+    let length_size = avc1.avcc.length_size_minus_one + 1;
+
+    let thread = std::thread::Builder::new()
+        .name(format!("mp4 reader {input_id}"))
+        .spawn(move || {
+            reader_thread(
+                sps_and_pps_payload,
+                reader,
+                sender,
+                stop_thread,
+                length_size,
+                TrackInfo {
+                    sample_count,
+                    timescale,
+                    track_id,
+                },
```

Comment:
yeah, but no other way to hoist the clousure up. can't pass the `track` from which these variables come from, because it borrows the `Mp4Reader` which I also pass in here.